### ISSUE TITLE
Unexpected signed integer or fixed point type: UNREACHABLE executed at ASTContext.cpp:9789!

### ISSUE THREAD

**Issue by e57aef97-3df7-40d0-9af8-3fbb13e07541** [llvm/llvm-project#46093]
|  |  |
| --- | --- |
| Bugzilla Link | [46748](https://llvm.org/bz46748) |
| Version | 11.0 |
| OS | Linux |
| Attachments | [Preprocessed source file](https://user-images.githubusercontent.com/92601330/143760983-1177936b-6a2c-42b6-8989-e82a11c3ef14.gz), [Associated run script](https://user-images.githubusercontent.com/92601330/143760984-66d0ccc7-1dec-4746-9d28-ef0c34d2b9ec.gz), [Error Output and Stacktrace](https://user-images.githubusercontent.com/92601330/143760985-a60cda04-da2f-4637-856b-6883523d387f.gz) |
| CC | @zygoloid |

## Extended Description 

Compiling the following (reduced) file with clang 11.0 (compiled from trunk) causes clang to crash with the following message:

$ cat bug.cpp
void a() {
  wchar_t *b;
  b = (char*)v
}

Error (full error output attached):
Unexpected signed integer or fixed point type
UNREACHABLE executed at /home/user/llvm-project/clang/lib/AST/ASTContext.cpp:9789!

Note: Clang version 10.0.0 just reports a compile error and doesn't crash.

Additional Information:

$ clang++ --version
clang version 11.0.0 (https://github.com/llvm/llvm-project.git e73d0b5719966ddbeff7a3da70a3cb782c3733ed)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /home/user/llvm-project/build/bin

$ cat /etc/os-release
NAME="openSUSE Tumbleweed"
# VERSION="20200710"

**Comment by k-mana** [comment]
The reproduction code is wrong.
The attached reproduction code is below.
```
void a() {
  wchar_t *b;
  b = (char*)c
}
```
### backtrace
```
Unexpected signed integer or fixed point type
UNREACHABLE executed at /home/k-mana/llvm-project-11.0.0/clang/lib/AST/ASTContext.cpp:9902!
PLEASE submit a bug report to https://bugs.llvm.org/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.      Program arguments: /opt/llvm-11/bin/clang -c t.cpp
1.      t.cpp:4:1: current parser token '}'
2.      t.cpp:1:10: parsing function body 'a'
3.      t.cpp:1:10: in compound statement ('{}')
 #0 0x0000000004ee6f1f llvm::sys::PrintStackTrace(llvm::raw_ostream&) /home/k-mana/llvm-project-11.0.0/llvm/lib/Support/Unix/Signals.inc:564:13
 #1 0x0000000004ee5170 llvm::sys::RunSignalHandlers() /home/k-mana/llvm-project-11.0.0/llvm/lib/Support/Signals.cpp:69:18
 #2 0x0000000004e6ee03 (anonymous namespace)::CrashRecoveryContextImpl::HandleCrash(int, unsigned long) /home/k-mana/llvm-project-11.0.0/llvm/lib/Support/CrashRecoveryContext.cpp:77:5
 #3 0x0000000004e6ef56 CrashRecoverySignalHandler(int) /home/k-mana/llvm-project-11.0.0/llvm/lib/Support/CrashRecoveryContext.cpp:383:1
 #4 0x00007fd81c3e2520 (/lib/x86_64-linux-gnu/libc.so.6+0x42520)
 #5 0x00007fd81c436a7c pthread_kill (/lib/x86_64-linux-gnu/libc.so.6+0x96a7c)
 #6 0x00007fd81c3e2476 raise (/lib/x86_64-linux-gnu/libc.so.6+0x42476)
 #7 0x00007fd81c3c87f3 abort (/lib/x86_64-linux-gnu/libc.so.6+0x287f3)
 #8 0x0000000004e73981 (/opt/llvm-11/bin/clang+0x4e73981)
 #9 0x0000000006f95558 llvm::isa_impl_cl<clang::EnumType, clang::Type const*>::doit(clang::Type const*) /home/k-mana/llvm-project-11.0.0/llvm/include/llvm/Support/Casting.h:104:5
#10 0x0000000006f95558 llvm::isa_impl_wrap<clang::EnumType, clang::Type const*, clang::Type const*>::doit(clang::Type const* const&) /home/k-mana/llvm-project-11.0.0/llvm/include/llvm/Support/Casting.h:131:12
#11 0x0000000006f95558 llvm::isa_impl_wrap<clang::EnumType, clang::QualType const, clang::Type const*>::doit(clang::QualType const&) /home/k-mana/llvm-project-11.0.0/llvm/include/llvm/Support/Casting.h:121:12
#12 0x0000000006f95558 bool llvm::isa<clang::EnumType, clang::QualType>(clang::QualType const&) /home/k-mana/llvm-project-11.0.0/llvm/include/llvm/Support/Casting.h:142:10
#13 0x0000000006f95558 std::enable_if<!(is_simple_type<clang::QualType>::value), llvm::cast_retty<clang::EnumType, clang::QualType const>::ret_type>::type llvm::dyn_cast<clang::EnumType, clang::QualType>(clang::QualType const&) /home/k-mana/llvm-project-11.0.0/llvm/include/llvm/Support/Casting.h:335:10
#14 0x0000000006f95558 clang::EnumType const* clang::Type::getAs<clang::EnumType>() const /home/k-mana/llvm-project-11.0.0/build/tools/clang/include/clang/AST/TypeNodes.inc:85:1
#15 0x0000000006f95558 clang::ASTContext::getCorrespondingUnsignedType(clang::QualType) const /home/k-mana/llvm-project-11.0.0/clang/lib/AST/ASTContext.cpp:9859:28
#16 0x00000000069ee4db checkPointerTypesForAssignment(clang::Sema&, clang::QualType, clang::QualType) /home/k-mana/llvm-project-11.0.0/clang/lib/Sema/SemaExpr.cpp:8721:26
#17 0x00000000069ee4db clang::Sema::CheckAssignmentConstraints(clang::QualType, clang::ActionResult<clang::Expr*, true>&, clang::CastKind&, bool) /home/k-mana/llvm-project-11.0.0/clang/lib/Sema/SemaExpr.cpp:9034:14
#18 0x00000000069ecd3a clang::Sema::CheckAssignmentConstraints(clang::SourceLocation, clang::QualType, clang::QualType) /home/k-mana/llvm-project-11.0.0/clang/lib/Sema/SemaExpr.cpp:8879:3
#19 0x0000000006ab7691 clang::Sema::PerformImplicitConversion(clang::Expr*, clang::QualType, clang::ImplicitConversionSequence const&, clang::Sema::AssignmentAction, clang::Sema::CheckedConversionKind) /home/k-mana/llvm-project-11.0.0/clang/lib/Sema/SemaExprCXX.cpp:4016:16
#20 0x0000000006c783b9 clang::Sema::PerformImplicitConversion(clang::Expr*, clang::QualType, clang::Sema::AssignmentAction, bool, clang::ImplicitConversionSequence&) /home/k-mana/llvm-project-11.0.0/clang/lib/Sema/SemaOverload.cpp:1523:1
#21 0x0000000006c78219 clang::Sema::PerformImplicitConversion(clang::Expr*, clang::QualType, clang::Sema::AssignmentAction, bool) /home/k-mana/llvm-project-11.0.0/clang/lib/Sema/SemaOverload.cpp:1498:10
#22 0x00000000069ef6d9 clang::Sema::CheckSingleAssignmentConstraints(clang::QualType, clang::ActionResult<clang::Expr*, true>&, bool, bool, bool) /home/k-mana/llvm-project-11.0.0/clang/lib/Sema/SemaExpr.cpp:9323:13
#23 0x00000000069fea26 clang::Sema::CheckAssignmentOperands(clang::Expr*, clang::ActionResult<clang::Expr*, true>&, clang::SourceLocation, clang::QualType) /home/k-mana/llvm-project-11.0.0/clang/lib/Sema/SemaExpr.cpp:12759:9
#24 0x00000000069db033 clang::Sema::getLangOpts() const /home/k-mana/llvm-project-11.0.0/clang/include/clang/Sema/Sema.h:1430:51
#25 0x00000000069db033 clang::Sema::CreateBuiltinBinOp(clang::SourceLocation, clang::BinaryOperatorKind, clang::Expr*, clang::Expr*) /home/k-mana/llvm-project-11.0.0/clang/lib/Sema/SemaExpr.cpp:13744:9
#26 0x00000000069c3078 clang::Sema::ActOnBinOp(clang::Scope*, clang::SourceLocation, clang::tok::TokenKind, clang::Expr*, clang::Expr*) /home/k-mana/llvm-project-11.0.0/clang/lib/Sema/SemaExpr.cpp:14170:1
#27 0x000000000662b559 clang::Parser::ParseRHSOfBinaryExpression(clang::ActionResult<clang::Expr*, true>, clang::prec::Level) /home/k-mana/llvm-project-11.0.0/clang/lib/Parse/ParseExpr.cpp:629:21
#28 0x000000000662a669 clang::Parser::ParseAssignmentExpression(clang::Parser::TypeCastState) /home/k-mana/llvm-project-11.0.0/clang/lib/Parse/ParseExpr.cpp:0:10
#29 0x000000000662a549 clang::Parser::ParseExpression(clang::Parser::TypeCastState) /home/k-mana/llvm-project-11.0.0/clang/lib/Parse/ParseExpr.cpp:125:10
#30 0x000000000664f375 clang::Parser::ParseExprStatement(clang::Parser::ParsedStmtContext) /home/k-mana/llvm-project-11.0.0/clang/lib/Parse/ParseStmt.cpp:440:19
#31 0x000000000664dbaa clang::Parser::ParseStatementOrDeclarationAfterAttributes(llvm::SmallVector<clang::Stmt*, 32u>&, clang::Parser::ParsedStmtContext, clang::SourceLocation*, clang::Parser::ParsedAttributesWithRange&) /home/k-mana/llvm-project-11.0.0/clang/lib/Parse/ParseStmt.cpp:234:12
#32 0x000000000664d622 clang::Parser::ParseStatementOrDeclaration(llvm::SmallVector<clang::Stmt*, 32u>&, clang::Parser::ParsedStmtContext, clang::SourceLocation*) /home/k-mana/llvm-project-11.0.0/clang/lib/Parse/ParseStmt.cpp:106:20
#33 0x0000000006655932 clang::Parser::ParseCompoundStatementBody(bool) /home/k-mana/llvm-project-11.0.0/clang/lib/Parse/ParseStmt.cpp:1089:11
#34 0x0000000006656830 clang::Parser::ParseFunctionStatementBody(clang::Decl*, clang::Parser::ParseScope&) /home/k-mana/llvm-project-11.0.0/clang/lib/Parse/ParseStmt.cpp:2237:21
#35 0x00000000065d8243 clang::Parser::ParseFunctionDefinition(clang::ParsingDeclarator&, clang::Parser::ParsedTemplateInfo const&, clang::Parser::LateParsedAttrList*) /home/k-mana/llvm-project-11.0.0/clang/lib/Parse/Parser.cpp:0:0
#36 0x00000000066033d6 clang::Parser::ParseDeclGroup(clang::ParsingDeclSpec&, clang::DeclaratorContext, clang::SourceLocation*, clang::Parser::ForRangeInit*) /home/k-mana/llvm-project-11.0.0/clang/lib/Parse/ParseDecl.cpp:1926:18
#37 0x00000000065d729c clang::Parser::ParseDeclOrFunctionDefInternal(clang::Parser::ParsedAttributesWithRange&, clang::ParsingDeclSpec&, clang::AccessSpecifier) /home/k-mana/llvm-project-11.0.0/clang/lib/Parse/Parser.cpp:0:10
#38 0x00000000065d6c4c clang::Parser::ParseDeclarationOrFunctionDefinition(clang::Parser::ParsedAttributesWithRange&, clang::ParsingDeclSpec*, clang::AccessSpecifier) /home/k-mana/llvm-project-11.0.0/clang/lib/Parse/Parser.cpp:1115:12
#39 0x00000000065d5ba4 clang::Parser::ParseExternalDeclaration(clang::Parser::ParsedAttributesWithRange&, clang::ParsingDeclSpec*) /home/k-mana/llvm-project-11.0.0/clang/lib/Parse/Parser.cpp:0:12
#40 0x00000000065d3d8d clang::Parser::ParseTopLevelDecl(clang::OpaquePtr<clang::DeclGroupRef>&, bool) /home/k-mana/llvm-project-11.0.0/clang/lib/Parse/Parser.cpp:683:10
#41 0x00000000065d3667 clang::Parser::ParseFirstTopLevelDecl(clang::OpaquePtr<clang::DeclGroupRef>&) /home/k-mana/llvm-project-11.0.0/clang/lib/Parse/Parser.cpp:555:26
#42 0x00000000065cf39e clang::ParseAST(clang::Sema&, bool, bool) /home/k-mana/llvm-project-11.0.0/clang/lib/Parse/ParseAST.cpp:0:25
#43 0x0000000005706714 clang::FrontendAction::Execute() /home/k-mana/llvm-project-11.0.0/clang/lib/Frontend/FrontendAction.cpp:954:10
#44 0x000000000567f763 llvm::Error::getPtr() const /home/k-mana/llvm-project-11.0.0/llvm/include/llvm/Support/Error.h:274:42
#45 0x000000000567f763 llvm::Error::operator bool() /home/k-mana/llvm-project-11.0.0/llvm/include/llvm/Support/Error.h:236:16
#46 0x000000000567f763 clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) /home/k-mana/llvm-project-11.0.0/clang/lib/Frontend/CompilerInstance.cpp:984:23
#47 0x00000000057b1668 clang::ExecuteCompilerInvocation(clang::CompilerInstance*) /home/k-mana/llvm-project-11.0.0/clang/lib/FrontendTool/ExecuteCompilerInvocation.cpp:278:25
#48 0x0000000003187f15 cc1_main(llvm::ArrayRef<char const*>, char const*, void*) /home/k-mana/llvm-project-11.0.0/clang/tools/driver/cc1_main.cpp:240:15
#49 0x00000000031860c9 ExecuteCC1Tool(llvm::SmallVectorImpl<char const*>&) /home/k-mana/llvm-project-11.0.0/clang/tools/driver/driver.cpp:330:12
#50 0x00000000055850b2 clang::driver::CC1Command::Execute(llvm::ArrayRef<llvm::Optional<llvm::StringRef> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, bool*) const::$_1::operator()() const /home/k-mana/llvm-project-11.0.0/clang/lib/Driver/Job.cpp:400:30
#51 0x00000000055850b2 void llvm::function_ref<void ()>::callback_fn<clang::driver::CC1Command::Execute(llvm::ArrayRef<llvm::Optional<llvm::StringRef> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, bool*) const::$_1>(long) /home/k-mana/llvm-project-11.0.0/llvm/include/llvm/ADT/STLExtras.h:185:12
#52 0x0000000004e6ed17 llvm::function_ref<void ()>::operator()() const /home/k-mana/llvm-project-11.0.0/llvm/include/llvm/ADT/STLExtras.h:0:12
#53 0x0000000004e6ed17 llvm::CrashRecoveryContext::RunSafely(llvm::function_ref<void ()>) /home/k-mana/llvm-project-11.0.0/llvm/lib/Support/CrashRecoveryContext.cpp:418:3
#54 0x000000000558457d clang::driver::CC1Command::Execute(llvm::ArrayRef<llvm::Optional<llvm::StringRef> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, bool*) const /home/k-mana/llvm-project-11.0.0/clang/lib/Driver/Job.cpp:400:7
#55 0x0000000005558f45 clang::driver::Compilation::ExecuteCommand(clang::driver::Command const&, clang::driver::Command const*&) const /home/k-mana/llvm-project-11.0.0/clang/lib/Driver/Compilation.cpp:195:15
#56 0x0000000005559327 clang::driver::Compilation::ExecuteJobs(clang::driver::JobList const&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*> >&) const /home/k-mana/llvm-project-11.0.0/clang/lib/Driver/Compilation.cpp:246:13
#57 0x000000000556c7ac llvm::SmallVectorBase<unsigned int>::empty() const /home/k-mana/llvm-project-11.0.0/llvm/include/llvm/ADT/SmallVector.h:69:47
#58 0x000000000556c7ac clang::driver::Driver::ExecuteCompilation(clang::driver::Compilation&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*> >&) /home/k-mana/llvm-project-11.0.0/clang/lib/Driver/Driver.cpp:1509:23
#59 0x00000000031859bf main /home/k-mana/llvm-project-11.0.0/clang/tools/driver/driver.cpp:502:21
#60 0x00007fd81c3c9d90 (/lib/x86_64-linux-gnu/libc.so.6+0x29d90)
#61 0x00007fd81c3c9e40 __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x29e40)
#62 0x0000000003182f85 _start (/opt/llvm-11/bin/clang+0x3182f85)
clang-11: error: clang frontend command failed due to signal (use -v to see invocation)
clang version 11.0.0
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /opt/llvm-11/bin
clang-11: note: diagnostic msg:
********************

PLEASE ATTACH THE FOLLOWING FILES TO THE BUG REPORT:
Preprocessed source(s) and associated run script(s) are located at:
clang-11: note: diagnostic msg: /tmp/t-18ea3e.cpp
clang-11: note: diagnostic msg: /tmp/t-18ea3e.sh
clang-11: note: diagnostic msg:

********************
```

**Comment by k-mana** [comment]
This no longer reproduces on trunk: https://godbolt.org/z/v1YhErbaP
This has been fixed: 95ce9fb
